### PR TITLE
Omg 314 - document interaction between MVP and MoreVP

### PR DESCRIPTION
**Pull request by paulperegud**

### CODE REVIEW

File: docs/mvp_morevp_interaction.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+* `game:` solution is an interactive game, possibly with new functions exposed by the contract.
+
+### Possible scenarios
+
+1. SE on input -> IFE  
+    => (IMPLEMENTED) contract: IFE challenges SE (~ +30k gas)  
+    NOTE: can do that even if I remembered standardExitId(hash(tx), oindex) on SE, because I am getting the input txs all over again in IFE
+
+2. SE on input, finalized -> IFE  
+    => (IMPLEMENTED) contract: IFE is "as if non-canonical" and can't be canonized; finalized input is marked as done (req. additional flag on IFE)
```
Comment (pthomalla) [inline docs/mvp_morevp_interaction.md]: "finalized input is marked as done" <- imprecise sentence

File: docs/mvp_morevp_interaction.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+3. IFE -> SE on input  
+    => (IMPLEMENTED) game: standard SE challenge using IFE tx as spend
+
+4. IFE, finalized input -> SE on input
```
Comment (pthomalla) [inline docs/mvp_morevp_interaction.md]: redundant "input"

File: docs/mvp_morevp_interaction.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+7. IFE -> SE on output  
+    => (IMPLEMENTED) contract: take tx from SE, find tx in IFEs (compute hash), block future piggybacks on output (gas pending)  
+    NOTE: +5k gas (after constantinople)
+
+8. IFE -> SE on output -> SE finalized -> piggyback output  
+   ! not possible
+
+8. IFE -> piggyback output -> SE on output  
+    => (IMPLEMENTED) contract: check for IFE & piggyback (compute hash)
```
Comment (pthomalla) [inline docs/mvp_morevp_interaction.md]: more detail

File: docs/mvp_morevp_interaction.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+8. IFE -> piggyback output -> SE on output  
+    => (IMPLEMENTED) contract: check for IFE & piggyback (compute hash)  
+    NOTE: +18k gas on startSE
+
+10. IFE -> piggyback output -> finalized -> SE on output  
+    ! same as (8.)
+
+### Exploring standardExitId(hash(tx), oindex)
+
+Q: does this affect standard exit and standard challenge?
```
Comment (pthomalla) [inline docs/mvp_morevp_interaction.md]: add blank lines

File: docs/mvp_morevp_interaction.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+### Exploring standardExitId(hash(tx), oindex)
+
+Q: does this affect standard exit and standard challenge?  
+A: standard challenge addresses inputs by utxo_pos; this change breaks this link  
+Q: can this be fixed?  
+A1: we can store utxo_pos as one of the fields of Exit structure (+20k gas); challenger will deliver exit_id (negligible cost).  
+Q: Could possibly drop utxo pos in favor of using txhash everywhere?  
+A: No, utxo pos is critical for MVP.  
+Q: Possible collision: inFlightExitId(hash(tx)) == standardExitId(hash(tx), 0)
```
Comment (pthomalla) [inline docs/mvp_morevp_interaction.md]: why we use standardExitId?

Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: Added an explanation of why this way of computing the value is introduced.

File: docs/morevp.md [hunk 6]
```diff
@@ -203,6 +203,9 @@ Otherwise, the exiting transaction is determined to be non-canonical and the cha
 Note that this challenge means it’s possible for an honest user to lose `exit bond` as they might not be aware their transaction is non-canonical.
 We address this attack vector and several mitigations in detail later.
 
+It might also be the case that MoreVP exit is opened where some of the inputs where referenced in MVP exit and those MVP exits were finalized.
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/morevp.md]: done

File: docs/morevp.md [hunk 7]
```diff
@@ -235,10 +238,26 @@ Any inputs or outputs paid out should be saved in the contract so that any futur
 #### Combining with Plasma MVP Exit Protocol
 
 The MoreVP protocol can be combined with the Plasma MVP protocol in a way that simultaneously preserves the integrity of exits and minimizes gas cost.
-Owners of outputs on the Plasma chain should be able to start an exit via either mechanism, but not both.
-Implementers can check this whenever an MVP exit is being submitted or whenever a MoreVP exit is being piggybacked.
-Although the two protocols use different determinations for exit priority, we still need a total ordering on exits.
+Although the two protocols use different determinations for exit priority, total ordering on exits is still needed.
 Therefore, every exit, no matter the protocol used, must be included in the same priority queue for processing.
+Honest user which enjoys data availability should be able to ignore MoreVP exits initiated by other parties.
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/morevp.md]: done

File: docs/mvp_morevp_interaction.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
+MVP vs MoreVP
+==
+
+Since both MVP standard exit and MoreVP in-flight exit provide an exit opportunity, it is possible to effectively double-spending. To prevent such possiblity, and to preserve the rule that honest owner can ignore IFE provided that he enjoys data availability, following scenarios are evaluated.
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md (hunk 1 above)
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    NOTE: costs: +20k on startSE; low cost on PB check  
+    NOTE: make revertable, so we can implement alt game solution later  
+    => (alt) (if data availability) game: piggyback is challenged with new challenge type  
+    => (alt) (if no data avail) do nothing: IFE output exits, SE exits after all mass exits  
+    NOTE: why can't we purely in contract (as in 7.) here? B/c "important" info (tx hash) is delivered too early and forgotten.  
+    But if we do standardExitId(...) then ok  
+    NOTE: as rule, we always block whatever is started later (piggyback here, SE in 8./9.)
+
+6. SE on output, finalized -> IFE -> piggyback output  
+    => (IMPLEMENTED) contract: standardExitId(hash(tx), oindex)
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md (hunk 2 above)
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+* `IFE` - in-flight exit, active unless stated otherwise;
+* `SE` - standard exit, active unless stated otherwise;
+* `SE on (in|out)put` - standard exit on utxo mentioned in in-flight exit as input or output;
+* `->` denotes order of events;
+* `=>` denotes proposed solution / action;
+* `contract:` solution is contained to change in the contract;
+* `game:` solution is an interactive game, possibly with new functions exposed by the contract.
+
+### Possible scenarios
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+Q: Ordering in the priority queue?  
+A: Priority is determined by most significant 64 bits out of 256, which are a timestamp anyway  
+
+Q: Can operator exploit his ability to include tx multiple times combined with this change?  
+A: I don't see a way to do it yet.  
+
+Q: When IFE/piggyback on output is added, what is the cost of checking if standard exits were performed from output(s)?  
+A: Number of outputs is known on both IFE and PB; single SLOAD per output?  
+
+### Anatomy of exit id and priority value
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: Moved to function docs in the contract

File: docs/mvp_morevp_interaction.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+   ! not possible
+
+9. IFE -> piggyback output -> SE on output  
+    => (IMPLEMENTED) contract: when SE happens, check for IFE & piggyback existence  
+    NOTE: +18k gas on startSE
+
+10. IFE -> piggyback output -> finalized -> SE on output  
+    ! same as (8.)
+
+### Exploring `standardExitId(hash(tx), oindex)`
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    => (IMPLEMENTED) contract: introduce `standardExitId(hash(tx), oindex)`, check if SE exists, if so treat IFE is "as if non-canonical" and can't be canonized; input is marked as finalized with regard to piggybacking. Add additional flag to IFE which will prevent it from taking part in canonicity game.
+
+3. IFE -> SE on input  
+    => (IMPLEMENTED) game: standard SE challenge using IFE tx as spend
+
+4. IFE, finalized input -> SE on input  
+    => (IMPLEMENTED) same as (3.)
+
+5. SE on output -> IFE -> piggyback output  
+    => (IMPLEMENTED) contract: `standardExitId(hash(tx), oindex)`. This allows to find SE when PB is performed. Need to store utxo_pos in Exit struct during `startStandardExit`, though.
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+6. SE on output, finalized -> IFE -> piggyback output  
+    => (IMPLEMENTED) contract: introduce `standardExitId(hash(tx), oindex)`; on PB compute SE id, check if it exists  
+    => (alternative) exit game as in 5. Piggyback is challenged (take bond)
+
+7. IFE -> SE on output  
+    => (IMPLEMENTED) contract: take tx from SE, find tx in IFEs, block future piggybacks on output  
+    NOTE: +5k gas (after constantinople)
+
+8. IFE -> SE on output -> SE finalized -> piggyback  
+   ! not possible
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
+MVP vs MoreVP
+==
+
+Since both MVP exit (aka standard exit aka SE) and MoreVP exit (aka in-flight exit aka IFE) provide an exit opportunity, it is possible to effectively double-spending. To prevent such possiblity, and to preserve the rule that honest owner can ignore IFE provided that he enjoys data availability, following scenarios are evaluated.
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
+MVP vs MoreVP
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/morevp.md [hunk 17]
```diff
@@ -235,10 +238,26 @@ Any inputs or outputs paid out should be saved in the contract so that any futur
...
-Owners of outputs on the Plasma chain should be able to start an exit via either mechanism, but not both.
-Implementers can check this whenever an MVP exit is being submitted or whenever a MoreVP exit is being piggybacked.
-Although the two protocols use different determinations for exit priority, we still need a total ordering on exits.
+Although the two protocols use different determinations for exit priority, total ordering on exits is still needed.
 Therefore, every exit, no matter the protocol used, must be included in the same priority queue for processing.
+Honest user which enjoys data availability should be able to ignore MoreVP exits that involve their outputs.
+Owners of outputs on the Plasma chain should be able to start an exit via either mechanism, but not both.
+To guarantee that money can't be double-spend via those two mechanisms, two approaches are possible.
+
+Option (1).
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/morevp.md]: done

File: docs/morevp.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/morevp.md]: done

File: docs/morevp.md [hunk 18]
```diff
@@ -235,10 +238,26 @@ Any inputs or outputs paid out should be saved in the contract so that any futur
...
+Contract needs to check if other type of exit exists for particular output when MVP exit is being submitted and it checks if MVP exit is in progress / was finalized when MoreVP exit is being added.
+In first case new exit is blocked. In second case - MoreVP exit is marked as one which can be exited only from inputs, and problematic inputs are marked as spent for piggybacking purposes.
+To make such checks possible, both types of exits need to use transaction hash as an exit id.
+No additional interactive games arise from the fact of coexistence of MVP and MoreVP protocols.
+
+Option (2).
+To reduce gas costs for honest participants, new types of challenges needs to be introduced.
+Piggybacks on outputs should be challenged by MVP exits and vice-versa.
+MVP exits on UTXO seen as the input of a MoreVP tx exit can be challenged using tx body.
+Canonicity of MoreVP exit can be removed by pointing contract to finalized MVP exit from MoreVP inputs, marking particular input as spent.
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/morevp.md]: I've alicebobbified all of the scenarios in docs with details and added a link.

File: docs/mvp_morevp_interaction.md [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+
+1. SE on input -> IFE  
+    => (IMPLEMENTED) contract: IFE challenges SE (~ +30k gas)  
+    NOTE: can do that even if I remembered `standardExitId(hash(tx), oindex)` on SE, because I am getting the input txs all over again in IFE
+
+2. SE on input, finalized -> IFE  
+    => (IMPLEMENTED) contract: introduce `standardExitId(hash(tx), oindex)`, check if SE exists, if so treat IFE is "as if non-canonical" and can't be canonized; input is marked as finalized with regard to piggybacking. Add additional flag to IFE which will prevent it from taking part in canonicity game.
+
+3. IFE -> SE on input  
+    => (IMPLEMENTED) game: standard SE challenge using IFE tx as spend
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: I've intended to communicate that challenge type is already in the contract. My bad.

File: docs/mvp_morevp_interaction.md (hunk 4 above)
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md (hunk 3 above)
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

File: docs/mvp_morevp_interaction.md [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+6. SE on output, finalized -> IFE -> piggyback output  
+    => (IMPLEMENTED) contract: standardExitId(hash(tx), oindex)  
+    => (alternative) exit game as in 5. Piggyback is challenged (take bond)
+
+7. IFE -> SE on output  
+    => (IMPLEMENTED) contract: take tx from SE, find tx in IFEs (compute hash), block future piggybacks on output (gas pending)  
+    NOTE: +5k gas (after constantinople)
+
+8. IFE -> SE on output -> SE finalized -> piggyback output
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: done

**Comment by paulperegud** [comment]
@pdobacz ready for followup CR

File: docs/mvp_morevp_interaction.md [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+8. Mallory submits IFE -> Mallory does SE on output -> Mallory waits for SE finalization -> Mallory piggybacks  
+   ! not possible, because IFE will finalize first because of use of "youngest input priority"
+
+9. Mallory submits IFE -> Mallory piggybacks output -> Mallory does SE on output  
+    => (IMPLEMENTED) contract: when SE happens, use tx body to compute in-flight exit id and check for IFE & piggyback existence  
+    NOTE: +18k gas on startSE
+
+10. Mallory submits IFE -> Mallory piggybacks output -> Mallory waits for finalization -> Mallory does SE on output  
+    ! same as (8.)
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply docs/mvp_morevp_interaction.md]: It was a case of bad reference. It was supposed to point at 9.

**Pull request merged**